samples/rust/rust_platdev.rs

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust platform device sample
```

Comment:
**That** is an interesting question! I definitely didn't want to add any duplicated Rust drivers to the canonical `drivers` tree. That might give people the mistaken impression that we are in the business of re-writing existing driver code, providing alternative implementations, etc. Which might evolve into a very thorny political issue !

I think we should make it *very* clear that this is an engineering example, which duplicates an existing driver only because it's convenient to do so for development, and definitely not because we wish encourage anyone to talk to `bcm2835-rng` through the Rust driver.

So maybe it would make sense to keep this in `samples/rust` ?

Reply:
> **That** is an interesting question! I definitely didn't want to add any duplicated Rust drivers to the canonical `drivers` tree. That might give people the mistaken impression that we are in the business of re-writing existing driver code, providing alternative implementations, etc. Which might evolve into a very thorny political issue !

No, it will be up to maintainers of particular drivers and subsystems.

Some maintainers will want nothing to be written in Rust, others may accept only new drivers in Rust, others may want to keep parallel implementations for a while, others may actually drop the C version as soon as possible, etc.

In other words, we will definitely have rewrites and alternative implementations happening (binder is actually such a case), but it is up to the maintainers to decide, not us, and therefore it is not an issue we need to get involved on.

> I think we should make it _very_ clear that this is an engineering example, which duplicates an existing driver only because it's convenient to do so for development, and definitely not because we wish encourage anyone to talk to `bcm2835-rng` through the Rust driver.
>
> So maybe it would make sense to keep this in `samples/rust` ?

It depends on what one is aiming with a particular driver -- that is why I was asking what your plan was. In particular, because AFAIU this is only an ARM-only, right? If yes, that makes it a prime candidate to be an actual driver since we will not be losing arch support, and so the maintainers might be open to the idea of having it, even potentially replacing the C one after a while.

Reply:
My "plan": I saw that senior Linux people wanted to see Rust touch "real" hardware. I have a little bit of Linux driver experience, and figured I could share some of that, to provide a starting point for Rust-for-Linux. I'm sure my naive abstractions are not something we should present upstream - but sometimes it's useful to have a starting point, however imperfect, which a community could improve. It's also interesting to contribute to this fascinating project, and learn something in the process.

I do not wish to get involved in the politics of providing alternative drivers or replacements. I'm however open to placing the driver anywhere in the tree, as long as it's committed with a clear disclaimer stating "this is not intended as a replacement". The maintainer of the subsystem in question may choose to remove the disclaimer later. I believe that was talked about in the Google Meet a few days ago.

> this is only an ARM-only driver, right?

This `hwrng` block is also used in a [BCM6368](https://www.broadcom.com/products/broadband/xdsl/bcm6368), which is a 32-bit mips based ADSL2 transceiver chip, over ten years old now. Not sure how it compares against Raspberry Pi - we could ask the `BCM6368` maintainer, Florian Fainelli.